splice_machine/src/main/java/com/splicemachine/derby/impl/stats/StoreCostControllerImpl.java

```diff
@@ -360,8 +359,37 @@ public RowLocation newRowLocationTemplate() {
         return null;
     }
 
+    private StringDataValue trimKey(StringDataValue key) throws StandardException {
+        if (key == null || key.getString() == null) {
+            return key;
+        }
+        return key.ansiTrim(StringDataValue.TRAILING, new SQLChar(" "), null);
+    }
+
+    private StringDataValue padKey(StringDataValue key, int length) throws StandardException {
+        if (key == null || key.getString() == null) {
+            return key;
+        }
+        SQLChar newStopKey = new SQLChar(key.getString());
+        newStopKey.setWidth(length, -1, true);
+        return newStopKey;
+    }
+
     @Override
-    public double getSelectivity(boolean fromExprIndex, int columnNumber, DataValueDescriptor start, boolean includeStart, DataValueDescriptor stop, boolean includeStop, boolean useExtrapolation) {
+    public double getSelectivity(boolean fromExprIndex, int columnNumber, DataValueDescriptor start, boolean includeStart, DataValueDescriptor stop, boolean includeStop, boolean useExtrapolation) throws StandardException {
+        if (useDb2CompatibleVarchars && baseTableRow.getColumn(columnNumber).getTypeName().equals(TypeId.VARCHAR_NAME)) {
+            int varcharLength = ((SQLVarchar)baseTableRow.getColumn(columnNumber)).getSqlCharSize();
+            if (includeStart) {
+                start = trimKey((StringDataValue)start);
+            } else {
+                start = padKey((StringDataValue) start, varcharLength);
```

Comment:
I don't think includeStart means that the start key value should change.  It just means when comparing with stats, do we use a > or a >= comparison (from ColumnStatisticsImpl.java):

> start != null && !start.isNull() && start.compare(includeStart ? ORDER_OP_GREATERTHAN : ORDER_OP_GREATEROREQUALS,

It seems we should always trim the start and always pad the end.  What do you think?

Reply:
Hum, I'm not sure. For example, if `startKey == 'a    '` and should be included, then startKey should be `'a'`. However, if it should not be included, then it means that `'a'` to `'a    ' should also not be included, so we need to pad it to the highest value not to be included.

Reply:
> Hum, I'm not sure. For example, if `startKey == 'a '` and should be included, then startKey should be `'a'`. However, if it should not be included, then it means that `'a'` to `'a '` should also not be included, so we need to pad it to the highest value not to be included.

Ah, right.  Everything matches up to the longest startKey, so to compute greater than it has to be greater than the longest key, and likewise less than end key has to be smaller than the shortest possible match.  Thanks.